glib/Signal.cs

```diff
@@ -276,15 +276,23 @@ public void RemoveDelegate (Delegate d)
 			if (tref.Target == null)
 				return;
 
-			if (d.Method.IsDefined (typeof (ConnectBeforeAttribute), false)) {
-				tref.Target.BeforeSignals [name] = Delegate.Remove (tref.Target.BeforeSignals [name] as Delegate, d);
-				if (tref.Target.BeforeSignals [name] == null && before_closure != null) {
+			bool before = d.Method.IsDefined (typeof (ConnectBeforeAttribute), false);
+			Hashtable hash = before ? tref.Target.BeforeSignals : tref.Target.AfterSignals;
+
+			Delegate del = Delegate.Remove (hash [name] as Delegate, d);
+			if (del != null) {
+				hash [name] = del;
+				return;
+			}
+
+			hash.Remove (del);
```

Comment:
Should this line not be `hash.Remove (d)` ? If `d` was the last delegate then `del` will be null when we hit this line.